File: src/dart/test/wire_format_test.dart
// Decodes the fixtures in ../../fixtures — real serde output from the Rust
// types — with the generated Dart classes.
//
// The schemas prove the shape agrees; these prove the *encoding* does. A
// generated class can compile against a correct schema and still fail here:
// `#[serde(skip_serializing_if)]` omits keys the schema lists, defaults arrive
// as absence rather than as a value, and an enum's wire spelling is nothing the
// class declaration can check for itself.
//
// Regenerate the fixtures with:
//   cargo run --locked --example generate_fixtures

import 'dart:convert';
import 'dart:io';

import 'package:test/test.dart';
import 'package:zed_interfaces/zed_interfaces.dart';

Map<String, dynamic> fixture(String name) =>
    jsonDecode(File('../../fixtures/$name.json').readAsStringSync()) as Map<String, dynamic>;

Map<String, dynamic> caseOf(String name, String key) =>
    fixture(name)[key] as Map<String, dynamic>;

/// Every key the server sent must survive a decode/encode round trip with an
/// equal value, at every depth. The reverse does not hold — Dart writes an
/// explicit null where serde omits the key entirely — so this is a one-way
/// containment check by design, and it has to recurse: a nested `AuditEntry`
/// gains those explicit nulls too, which a shallow `equals` would reject.
void expectContains(Object? original, Object? encoded, [String path = r'$']) {
  if (original is Map) {
    expect(encoded, isA<Map<String, dynamic>>(), reason: '$path should still be an object');
    final actual = encoded! as Map<String, dynamic>;
    for (final entry in original.entries) {
      expect(actual.containsKey(entry.key), isTrue, reason: '$path.${entry.key} was dropped');
      expectContains(entry.value, actual[entry.key], '$path.${entry.key}');
    }
    return;
  }
  if (original is List) {
    expect(encoded, isA<List<dynamic>>(), reason: '$path should still be a list');
    final actual = encoded! as List<dynamic>;
    expect(actual.length, original.length, reason: '$path changed length');
    for (var i = 0; i < original.length; i++) {
      expectContains(original[i], actual[i], '$path[$i]');
    }
    return;
  }
  expect(encoded, equals(original), reason: '$path changed value');
}

void expectRoundTrip(Map<String, dynamic> original, Map<String, dynamic> encoded) =>
    expectContains(original, encoded);

void main() {
  group('PackageMetadata', () {
    test('decodes a response with every optional field present', () {
      final json = caseOf('package-metadata', 'full');
      final decoded = PackageMetadata.fromJson(json);
      expect(decoded.org, 'acme');
      expect(decoded.vcs, Vcs.jj);
      expect(decoded.versionScheme, VersionScheme.calver);
      expect(decoded.tags, ['http', 'client']);
      expect(decoded.versions, ['2.0.0', '1.0.0']);
      expectRoundTrip(json, decoded.toJson());
    });

    test('decodes a response where serde omitted every optional key', () {
      // The minimal fixture has no `description`, `latest`, `tags` or
      // `version_scheme` at all — this is what `skip_serializing_if` produces,
      // and treating absence as an error would break every real response.
      final json = caseOf('package-metadata', 'minimal');
      expect(json.containsKey('version_scheme'), isFalse);
      final decoded = PackageMetadata.fromJson(json);
      expect(decoded.description, isNull);
      expect(decoded.latest, isNull);
      expect(decoded.tags, isNull);
      expect(decoded.versionScheme, isNull, reason: 'absent means "the default", not a value');
      expectRoundTrip(json, decoded.toJson());
    });
  });

  test('VersionMetadata decodes an integer size and a defaulted format', () {
    final json = caseOf('version-metadata', 'published');
    final decoded = VersionMetadata.fromJson(json);
    expect(decoded.size, 4096);
    expect(decoded.format, ArtifactFormat.tarGz);
    expect(decoded.yanked, isFalse);
    expect(decoded.vcsCommit, isNotNull);
    expectRoundTrip(json, decoded.toJson());
  });

  test('ApiError decodes the error envelope', () {
    final json = caseOf('api-error', 'not_found');
    final decoded = ApiError.fromJson(json);
    expect(decoded.code, 'not_found');
    expectRoundTrip(json, decoded.toJson());
  });

  test('list and search responses decode their shared PackageSummary', () {
    final list = caseOf('package-list-response', 'page');
    final decodedList = PackageListResponse.fromJson(list);
    expect(decodedList.total, 1);
    expect(decodedList.items.single.name, 'http-kit');
    expectRoundTrip(list, decodedList.toJson());

    final search = caseOf('search-response', 'hit');
    final decodedSearch = SearchResponse.fromJson(search);
    expect(decodedSearch.query, 'http');
    expect(decodedSearch.items.single.org, 'acme');
    expectRoundTrip(search, decodedSearch.toJson());
  });

  test('AuditLogResponse decodes a nested list with a nullable enum', () {
    final json = caseOf('audit-log-response', 'chain');
    final decoded = AuditLogResponse.fromJson(json);
    final entry = decoded.entries.single;
    expect(entry.actionKind, AuditAction.publish);
    expect(entry.prevHash, isNull, reason: 'the first entry has no predecessor');
    expect(entry.seq, 1);
    expectRoundTrip(json, decoded.toJson());
  });

  test('SyncChangeEvent decodes opaque JSON and a nested Hlc', () {
    final json = caseOf('sync-change-event', 'upsert');
    final decoded = SyncChangeEvent.fromJson(json);
    expect(decoded.op, SyncOp.upsert);
    expect(decoded.version.counter, 3);
    expect(decoded.atMs, 1754400000001);
    expect(decoded.row, {'org': 'acme'});
    expect(decoded.writeKey, isNull);
    expectRoundTrip(json, decoded.toJson());
  });

  test('enum wire spellings match what serde emits', () {
    // Each of these would be a silent decode failure in production if the
    // generated spelling drifted from the Rust variant's rename.
    final enums = fixture('enums');
    expect(SyncWriteMode.fromJson(enums['write_mode'] as String), SyncWriteMode.optimisticQueue);
    expect(SyncErrorPolicy.fromJson(enums['error_policy'] as String), SyncErrorPolicy.emitOnly);
    expect(
      SyncConflictResolution.fromJson(enums['conflict_resolution'] as String),
      SyncConflictResolution.serverWins,
    );
    expect(ArtifactFormat.fromJson(enums['artifact_format'] as String), ArtifactFormat.zip);
    expect(Vcs.fromJson(enums['vcs'] as String), Vcs.sapling);
    expect(VersionScheme.fromJson(enums['version_scheme'] as String), VersionScheme.opaque);
    expect(AuditAction.fromJson(enums['audit_action'] as String), AuditAction.orgClaim);
  });

  test('an unknown enum value fails loudly rather than decoding to a fallback', () {
    expect(() => Vcs.fromJson('perforce'), throwsFormatException);
  });

  test('the remaining request and response bodies decode', () {
    final misc = fixture('misc-responses');
    final publish = PublishResponse.fromJson(misc['publish'] as Map<String, dynamic>);
    expect(publish.version, '1.0.0');
    expectRoundTrip(misc['publish'] as Map<String, dynamic>, publish.toJson());

    final claimRequest = ClaimOrgRequest.fromJson(misc['claim_org_request'] as Map<String, dynamic>);
    expect(claimRequest.slug, 'acme');

    final claimResponse =
        ClaimOrgResponse.fromJson(misc['claim_org_response'] as Map<String, dynamic>);
    expect(claimResponse.created, isTrue);

    expect(YankRequest.fromJson(misc['yank_request'] as Map<String, dynamic>).yanked, isTrue);
    expect(YankResponse.fromJson(misc['yank_response'] as Map<String, dynamic>).name, 'http-kit');

    final integrity =
        AuditIntegrityResponse.fromJson(misc['audit_integrity'] as Map<String, dynamic>);
    expect(integrity.intact, isTrue);
    expect(integrity.entriesChecked, 1);
    expect(integrity.firstBadSeq, isNull);

    final request =
        SemanticSearchRequest.fromJson(misc['semantic_search_request'] as Map<String, dynamic>);
    expect(request.limit, 5);
    expect(request.embedding.length, 3);

    final response =
        SemanticSearchResponse.fromJson(misc['semantic_search_response'] as Map<String, dynamic>);
    expect(response.items.single.distance, closeTo(0.25, 1e-9));
  });
}

File: src/ts/wire-format.test.ts
// Checks the generated TypeScript types against the fixtures in ../../fixtures
// — real serde output from the Rust types.
//
//   node --test wire-format.test.ts
//
// TypeScript interfaces are erased at runtime, so a plain `as PackageMetadata`
// proves nothing on its own. Two things here are worth checking anyway, and
// both are real failure modes:
//
//   * the *compiler* sees each fixture assigned to its interface, so a missing
//     or misspelled required field fails `tsc` (that is the type-level half);
//   * the emitted `*_VALUES` arrays are runtime data, so every enum spelling
//     serde produces can be checked against them here (the runtime half).
//
// Regenerate the fixtures with:
//   cargo run --locked --example generate_fixtures

import assert from "node:assert/strict";
import fs from "node:fs";
import path from "node:path";
import { test } from "node:test";

import type {
  ApiError,
  AuditIntegrityResponse,
  AuditLogResponse,
  ClaimOrgRequest,
  ClaimOrgResponse,
  PackageListResponse,
  PackageMetadata,
  PublishResponse,
  SearchResponse,
  SemanticSearchRequest,
  SemanticSearchResponse,
  SyncChangeEvent,
  VersionMetadata,
  YankRequest,
  YankResponse,
} from "./index.ts";
import {
  ARTIFACT_FORMAT_VALUES,
  AUDIT_ACTION_VALUES,
  SYNC_CONFLICT_RESOLUTION_VALUES,
  SYNC_ERROR_POLICY_VALUES,
  SYNC_OP_VALUES,
  SYNC_WRITE_MODE_VALUES,
  VCS_VALUES,
  VERSION_SCHEME_VALUES,
} from "./index.ts";

const fixture = <T>(name: string): T =>
  JSON.parse(
    fs.readFileSync(path.join(import.meta.dirname, "../../fixtures", `${name}.json`), "utf8"),
  ) as T;

test("PackageMetadata: every optional field present", () => {
  const { full } = fixture<{ full: PackageMetadata }>("package-metadata");
  assert.equal(full.org, "acme");
  assert.equal(full.vcs, "jj");
  assert.equal(full.version_scheme, "calver");
  assert.deepEqual(full.tags, ["http", "client"]);
});

test("PackageMetadata: serde omitted every optional key", () => {
  // `skip_serializing_if` means these keys are *absent*, not null. The
  // interface marks them optional, which is what makes this assignment legal.
  const { minimal } = fixture<{ minimal: PackageMetadata }>("package-metadata");
  assert.equal("version_scheme" in minimal, false);
  assert.equal("description" in minimal, false);
  assert.equal(minimal.version_scheme ?? "semver", "semver");
});

test("VersionMetadata: numbers and defaulted fields", () => {
  const { published } = fixture<{ published: VersionMetadata }>("version-metadata");
  assert.equal(published.size, 4096);
  assert.equal(published.format ?? "tar.gz", "tar.gz");
  assert.equal(published.yanked ?? false, false);
});

test("responses that share PackageSummary decode consistently", () => {
  const { page } = fixture<{ page: PackageListResponse }>("package-list-response");
  const { hit } = fixture<{ hit: SearchResponse }>("search-response");
  assert.equal(page.total, 1);
  assert.equal(page.items[0]?.name, "http-kit");
  assert.equal(hit.query, "http");
  assert.equal(hit.items[0]?.org, "acme");
});

test("AuditLogResponse: nested entries with an optional enum", () => {
  const { chain } = fixture<{ chain: AuditLogResponse }>("audit-log-response");
  const entry = chain.entries[0];
  assert.ok(entry);
  assert.equal(entry.action_kind, "publish");
  assert.equal("prev_hash" in entry, false, "the first entry has no predecessor");
  assert.equal(entry.seq, 1);
});

test("SyncChangeEvent: opaque row and nested Hlc", () => {
  const { upsert } = fixture<{ upsert: SyncChangeEvent }>("sync-change-event");
  assert.equal(upsert.op, "upsert");
  assert.equal(upsert.version.counter, 3);
  assert.deepEqual(upsert.row, { org: "acme" });
  assert.ok(SYNC_OP_VALUES.includes(upsert.op));
});

test("every enum spelling serde emits is a member of the generated union", () => {
  // The union type is erased at runtime; the VALUES array is not. If a Rust
  // rename drifted from the generated spelling, this is where it surfaces —
  // a front end would otherwise narrow against a value that cannot occur.
  const enums = fixture<Record<string, string>>("enums");
  const pairs: ReadonlyArray<readonly [string, readonly string[]]> = [
    ["write_mode", SYNC_WRITE_MODE_VALUES],
    ["error_policy", SYNC_ERROR_POLICY_VALUES],
    ["conflict_resolution", SYNC_CONFLICT_RESOLUTION_VALUES],
    ["artifact_format", ARTIFACT_FORMAT_VALUES],
    ["vcs", VCS_VALUES],
    ["version_scheme", VERSION_SCHEME_VALUES],
    ["audit_action", AUDIT_ACTION_VALUES],
  ];
  for (const [key, values] of pairs) {
    const emitted = enums[key];
    assert.ok(emitted, `fixture is missing ${key}`);
    assert.ok(values.includes(emitted), `${key}="${emitted}" is not in [${values.join(", ")}]`);
  }
});

test("the remaining request and response bodies match their interfaces", () => {
  const misc = fixture<{
    publish: PublishResponse;
    claim_org_request: ClaimOrgRequest;
    claim_org_response: ClaimOrgResponse;
    yank_request: YankRequest;
    yank_response: YankResponse;
    audit_integrity: AuditIntegrityResponse;
    semantic_search_request: SemanticSearchRequest;
    semantic_search_response: SemanticSearchResponse;
  }>("misc-responses");

  assert.equal(misc.publish.version, "1.0.0");
  assert.equal(misc.claim_org_request.slug, "acme");
  assert.equal(misc.claim_org_response.created, true);
  assert.equal(misc.yank_request.yanked, true);
  assert.equal(misc.yank_response.name, "http-kit");
  assert.equal(misc.audit_integrity.intact, true);
  assert.equal(misc.audit_integrity.entries_checked, 1);
  assert.equal(misc.semantic_search_request.limit, 5);
  assert.equal(misc.semantic_search_request.embedding.length, 3);
  assert.equal(misc.semantic_search_response.items[0]?.distance, 0.25);
});

test("api-error", () => {
  const { not_found } = fixture<{ not_found: ApiError }>("api-error");
  assert.equal(not_found.code, "not_found");
});
